E2E: UI tests failing before they start

The e2e test workflow [here](https://github.com/hashicorp/nomad-e2e/blob/e6a28d9/.github/workflows/e2e-runner.yml#L289-L300) (private repo) runs this:

```
nomad namespace apply proxy
nomad job run ./e2e/ui/input/proxy.nomad
IP=$(nomad node status -json -verbose "$(nomad operator api '/v1/allocations?namespace=proxy' | jq -r '.[] | select(.JobID == "nomad-proxy") | .NodeID')" | jq -r '.Attributes."unique.platform.aws.public-ipv4"')
export NOMAD_ADDR="https://$IP:6464"
echo "running UI tests against $NOMAD_ADDR"
docker run -it --rm -v "$(pwd)"/e2e/ui:/src -w /src -e NOMAD_ADDR="$NOMAD_ADDR" -e NOMAD_TOKEN="$NOMAD_TOKEN" --ipc=host --net=host mcr.microsoft.com/playwright:focal bash script.sh
```

which fails like this:

```
Successfully applied namespace "proxy"!
==> 2023-11-16T20:39:21Z: Monitoring evaluation "382e659d"
    2023-11-16T20:39:21Z: Evaluation triggered by job "nomad-proxy"
    2023-11-16T20:39:22Z: Evaluation within deployment: "ecd7d0d0"
    2023-11-16T20:39:22Z: Allocation "8f33eca7" created: node "e170f7e1", group "proxy"
    2023-11-16T20:39:22Z: Evaluation status changed: "pending" -> "complete"
==> 2023-11-16T20:39:22Z: Evaluation "382e659d" finished with status "complete"
==> 2023-11-16T20:39:22Z: Monitoring deployment "ecd7d0d0"
    
2023-11-16T20:39:22Z
ID          = ecd7d0d0
Job ID      = nomad-proxy
Job Version = 0
Status      = running
Description = Deployment is running

Deployed
Task Group  Desired  Placed  Healthy  Unhealthy  Progress Deadline
proxy       1        1       0        0          2023-11-16T20:49:21Z
    
2023-11-16T20:39:38Z
ID          = ecd7d0d0
Job ID      = nomad-proxy
Job Version = 0
Status      = running
Description = Deployment is running

Deployed
Task Group  Desired  Placed  Healthy  Unhealthy  Progress Deadline
proxy       1        1       1        0          2023-11-16T20:49:38Z
    
2023-11-16T20:39:39Z
ID          = ecd7d0d0
Job ID      = nomad-proxy
Job Version = 0
Status      = successful
Description = Deployment completed successfully

Deployed
Task Group  Desired  Placed  Healthy  Unhealthy  Progress Deadline
proxy       1        1       1        0          2023-11-16T20:49:38Z
parse error: Invalid numeric literal at line 1, column 8
Prefix matched multiple nodes

ID                                    DC   Name              Class   Drain  Eligibility  Status
b035da60-b8ce-fbc1-4822-0dd4ee994fd6  dc1  EC2AMAZ-NHFTA7S   <none>  false  eligible     ready
e170f7e1-80d1-3e21-3ea6-7238fc996ecd  dc1  ip-172-31-87-242  <none>  false  eligible     ready
9336b433-b21c-184c-e2ca-6270dc88752f  dc2  ip-172-31-83-252  <none>  false  eligible     ready
87447de8-c5a5-2f3b-6448-025dfb192d74  dc1  ip-172-31-90-161  <none>  false  eligible     ready
41f03e06-2f49-5187-e8b7-f16fe73d75a2  dc2  ip-172-31-88-157  <none>  false  eligible     ready
running UI tests against https://:6464
the input device is not a TTY
Error: Process completed with exit code 1.
```

First gotta overcome the parse error trying to get an IP the proxy is running on, to use as a NOMAD_ADDR, then see if tests run okay.

--- a/e2e/ui/global-setup.js
+++ b/e2e/ui/global-setup.js
@@ -22,8 +22,9 @@ module.exports = async config => {
   const page = await context.newPage();
   await page.goto(NOMAD_ADDR+'/ui/settings/tokens');
   await page.fill('input[id="token-input"]', NOMAD_TOKEN);
-  await page.click('button:has-text("Set Token")', {strict: true});
+  await page.click('button:has-text("Sign in")', {strict: true});
 
-  await page.context().storageState({ path: 'storageState.json' });
+  const { storageState } = config.projects[0].use;
+  await page.context().storageState({ path: storageState });
   await browser.close();
 };

--- a/e2e/ui/input/proxy.nomad
+++ b/e2e/ui/input/proxy.nomad
@@ -19,6 +19,17 @@ job "nomad-proxy" {
       }
     }
 
+    service {
+      name     = "nomad-proxy"
+      port     = "www"
+      provider = "nomad"
+      check {
+        type     = "tcp"
+        interval = "1s"
+        timeout  = "2s"
+      }
+    }
+
     task "nginx" {
 
       driver = "docker"

--- a/e2e/ui/playwright.config.js
+++ b/e2e/ui/playwright.config.js
@@ -7,6 +7,8 @@
 // @ts-check
 const { devices } = require('@playwright/test');
 
+export const STORAGE_STATE = 'storageState.json';
+
 /** @type {import('@playwright/test').PlaywrightTestConfig} */
 const config = {
   forbidOnly: !!process.env.CI,
@@ -19,7 +21,10 @@ const config = {
   projects: [
     {
       name: 'chromium',
-      use: { ...devices['Desktop Chrome'] },
+      use: {
+        ...devices['Desktop Chrome'],
+        storageState: STORAGE_STATE,
+      },
     },
     // disabling firefox temporarily because the container doesn't
     // include it and so it tries to automatically install it and
@@ -31,7 +36,10 @@ const config = {
     // },
     {
       name: 'webkit',
-      use: { ...devices['Desktop Safari'] },
+      use: {
+        ...devices['Desktop Safari'],
+        storageState: STORAGE_STATE,
+      },
     },
   ],
 };

--- a/e2e/ui/run.sh
+++ b/e2e/ui/run.sh
@@ -2,8 +2,6 @@
 # Copyright (c) HashiCorp, Inc.
 # SPDX-License-Identifier: BUSL-1.1
 
-set -e
-
 help() {
     cat <<EOF
 Usage: run.sh [subcommand] [options] [--help]
@@ -22,6 +20,8 @@ Usage: run.sh [subcommand] [options] [--help]
          this so that we don't need to load a CA certificate into the browser.
          This reverse proxy uses a self-signed cert. Will print a new NOMAD_ADDR
          address for you to use for test runs.
+
+  ci     For use in CI: runs the proxy, then tests, then stops the proxy.
 
   Environment Variables:
   NOMAD_ADDR    Address of Nomad cluster or reverse proxy.
@@ -43,7 +43,9 @@ run_shell() {
 }
 
 run() {
-    exec docker run -it --rm \
+    local tty_args
+    [ -t 1 ] && tty_args='-it'
+    docker run $tty_args --rm \
            -v $(pwd):/src \
            -w /src \
            -e NOMAD_ADDR=$NOMAD_ADDR \
@@ -55,13 +57,48 @@ run() {
 }
 
 run_proxy() {
-    nomad namespace apply proxy
-    nomad job run "./input/proxy.nomad"
-    IP=$(nomad node status -json -verbose \
-          $(nomad operator api '/v1/allocations?namespace=proxy' | jq -r '.[] | select(.JobID == "nomad-proxy") | .NodeID') \
-        | jq -r '.Attributes."unique.platform.aws.public-ipv4"')
-    echo "NOMAD_ADDR=https://$IP:6464"
-    exit 0
+  # sending these outputs to stderr so that 'export NOMAD_ADDR='
+  # is the only stdout line, for scripts to eval.
+  nomad namespace apply proxy 1>&2 || return
+  nomad job run "./input/proxy.nomad" 1>&2
+  IP="$(_get_aws_ip)"
+  [ -n "$IP" ] || {
+    >&2 echo 'falling back to service IP'
+    IP="$(_get_svc_ip)"
+  }
+  [ -n "$IP" ] || {
+    >&2 echo 'unable to get an IP for nomad proxy...'
+    exit 1 # bad form to exit from a function, but this is essential (and eval'd)
+  }
+  echo "export NOMAD_ADDR=https://$IP:6464"
+}
+
+_get_aws_ip(){
+  aws_metadata_url="http://169.254.169.254/latest/meta-data"
+  nomad exec -namespace=proxy -job nomad-proxy \
+    curl -s "$aws_metadata_url/public-ipv4"
+}
+
+_get_svc_ip() {
+  nomad service info -namespace=proxy \
+    -t '{{ range . }}{{ .Address }}{{ end }}' \
+    nomad-proxy
+}
+
+stop_proxy() {
+  # make sure addr isn't still pointed at the proxy
+  export NOMAD_ADDR="${NOMAD_ADDR/6464/4646}"
+  nomad job stop -namespace=proxy nomad-proxy
+  nomad namespace delete proxy
+}
+
+run_ci() {
+  set -x
+  eval "$(run_proxy)"
+  run_tests
+  rc=$?
+  stop_proxy
+  exit $rc
 }
 
 opt="$1"
@@ -69,8 +106,8 @@ case $opt in
     help|--help|-h) help ;;
     proxy|--proxy) run_proxy ;;
     test|--test) shift ; run_tests "$@" ;;
-    shell) shift ; run_shell ;;
+    stop|--stop) stop_proxy ;;
+    ci|--ci) run_ci ;;
+    shell) shift ; run_shell "$@" ;;
     *) run_tests "$@" ;;
 esac
-
-run_tests

--- a/e2e/ui/run.sh
+++ b/e2e/ui/run.sh
@@ -2,6 +2,8 @@
 # Copyright (c) HashiCorp, Inc.
 # SPDX-License-Identifier: BUSL-1.1
 
+set -eu
+
 help() {
     cat <<EOF
 Usage: run.sh [subcommand] [options] [--help]
@@ -43,7 +45,7 @@ run_shell() {
 }
 
 run() {
-    local tty_args
+    local tty_args=''
     [ -t 1 ] && tty_args='-it'
     docker run $tty_args --rm \
            -v $(pwd):/src \
@@ -57,15 +59,17 @@ run() {
 }
 
 run_proxy() {
-  # sending these outputs to stderr so that 'export NOMAD_ADDR='
-  # is the only stdout line, for scripts to eval.
-  nomad namespace apply proxy 1>&2 || return
-  nomad job run "./input/proxy.nomad" 1>&2
+  # sending these outputs to stderr so that 'export NOMAD_ADDR=' is the
+  # only stdout line, for users to eval, or this script to write then source.
+  nomad namespace apply proxy 1>&2
+  nomad job run ./input/proxy.nomad 1>&2
+  set +e
   IP="$(_get_aws_ip)"
   [ -n "$IP" ] || {
     >&2 echo 'falling back to service IP'
     IP="$(_get_svc_ip)"
   }
+  set -e
   [ -n "$IP" ] || {
     >&2 echo 'unable to get an IP for nomad proxy...'
     exit 1 # bad form to exit from a function, but this is essential (and eval'd)
@@ -94,7 +98,8 @@ stop_proxy() {
 
 run_ci() {
   set -x
-  eval "$(run_proxy)"
+  run_proxy > /tmp/proxy_addr.env
+  source /tmp/proxy_addr.env
   run_tests
   rc=$?
   stop_proxy

--- a/e2e/ui/tests/example.spec.js
+++ b/e2e/ui/tests/example.spec.js
@@ -19,6 +19,6 @@ test('authenticated users can see their policies', async ({ page }) => {
   await expect(logo).toBeVisible();
 
 
-  const policies = page.locator('text=Policies')
+  const policies = page.getByText('Policies');
   await expect(policies).toBeVisible();
 });

